### PR TITLE
[Issue#606] Process.MainModule can throw NullReferenceException

### PR DESCRIPTION
**Pull request by pallavit**
Process library uses psapi!EnumProcessModules to get the complete module list and then uses the first module in the list to get the MainModule. This API has certain limitations and our implementation tries to idenitfy/work-around all these limitations by giving exceptions, calling the native API multiple times etc. However the API can still return null which is then wrapped in a ProcessModule causing NullReferenceException on usage.

As per msdn [EnumProcessModules](https://msdn.microsoft.com/en-us/library/windows/desktop/ms682631(v=vs.85).aspx) is not always guranteed to give a correct behavior.

    If the module list in the target process is corrupted or not yet initialized, or if
    the module list changes during the function call as a result of DLLs being loaded or unloaded,
    EnumProcessModules may fail or return incorrect information.

And so, it is not possible for us to always give the MainModule. As a result I have updated the implementation of Process.MainModule to return null (instead of wrapping the null in a ProcessModule). This pushes the onus on the callsite to check for this possibility. This is essentially a breaking change however it will break only those apps which earlier would have caused NullReferenceException anyway. And the developer could guard against this scenario with a simple null reference check code pattern. In fact in .Net 1.1 Process.MainModule did return null. The behavior got changed at some point between 1.1 and 2.0.

I also considered the following options
- Throw an exception - Since the behavior is caused due to things outside of developer control, the developer can't do anything to prevent this exception. It is unclear what type of exception should be thrown? And overall it seemed like an overkill to guard against a scenario, you could easily do with a Null check.
- Wrap the null but make all the properties return default values - This is the least breaking option, however this behavior is difficult to identify and in my opinion incorrect!

Interestingly I did not find any hits for the same behavior on desktop. I also tried running our open src test in loop for a day and the only time I could repro this was through ctrl+c.

### CODE REVIEW

**Comment by stephentoub** [comment]
The Process_MainModule test would need to be updated to check for the null return, too, right?  And [ActiveIssue] removed from it?

There's also a contract check at the beginning of GetModuleInfos that appears to be wrong, as it's verifying that the length is always >= 1. (Though to my knowledge these contracts aren't being used anyway.)

> This is essentially a breaking change however it will break only those apps which earlier would have caused NullReferenceException anyway

While that's not technically true (e.g. code could have been comparing MainModule to null but not accessing any of its members), I agree that's very unlikely in practice.

**Comment by pallavit** [comment]
@stephentoub Thanks for the feedback. I have updated the changes.

> The Process_MainModule test would need to be updated to check for the null return, too, right? And [ActiveIssue] removed from it?

Forgot to include the test changes in the commit.

> There's also a contract check at the beginning of GetModuleInfos that appears to be wrong, as it's verifying that the length is always >= 1. (Though to my knowledge these contracts aren't being used anyway.)

I have removed the contract package dependency from Process since we do not use contracts anyway.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 1]
```diff
@@ -13,7 +13,7 @@ namespace System.Diagnostics.ProcessTests
     public partial class ProcessTest : IDisposable
     {
         private const int WaitInMS = 100 * 1000; 
-        private const string CoreRunName = "corerun";
+        private const string CoreRunName = "corerun.exe";
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: Why this change?  This will break on Unix, where corerun doesn't have this suffix.

Comment (pallavit) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: MainModule returns the filename with the extension and so I updated this as it works on windows. I did not realize Unix won't work with this. I will make a commit that changes MainModule test instead. (In general, it looks like our CI system only does windows validation, how is one suppose to ensure that everything works cross-plat?)

Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: > it looks like our CI system only does windows validation, how is one suppose to ensure that everything works cross-plat?

That's about to change.  We now have xunit up and running on CoreCLR on Linux, and @mmitche just harnessed it in the CI system.   We'll be turning it on for PRs once a few issues are sorted out (e.g. at the moment we have a lot of failures because tests that are Windows-specific haven't yet been annotated as such).

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 2]
```diff
@@ -421,7 +427,7 @@ public void Process_InvalidPriorityClass()
         [Fact]
         public void ProcessProcessName()
         {
-            Assert.Equal(_process.ProcessName, CoreRunName, StringComparer.OrdinalIgnoreCase);
+            Assert.Equal(_process.ProcessName, Path.GetFileNameWithoutExtension(CoreRunName), StringComparer.OrdinalIgnoreCase);
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: Is this change still needed?

**Comment by stephentoub** [comment]
One minor nit, otherwise LGTM.

**Comment by stephentoub** [comment]
LGTM.  Thanks for fixing.
